ur_controllers/src/speed_scaling_state_controller.cpp

```diff
@@ -19,10 +19,187 @@
+    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
+  }
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SpeedScalingStateController::on_activate(const rclcpp_lifecycle::State& /*previous_state*/)
+{
+  if (!init_sensor_data())
+  {
+    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
+  }
+
+  last_publish_time_ = node_->now();
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SpeedScalingStateController::on_deactivate(const rclcpp_lifecycle::State& /*previous_state*/)
+{
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+}
+
+template <typename T>
+bool has_any_key(const std::unordered_map<std::string, T>& map, const std::vector<std::string>& keys)
+{
+  bool found_key = false;
+  for (const auto& key_item : map)
+  {
+    const auto& key = key_item.first;
+    if (std::find(keys.cbegin(), keys.cend(), key) != keys.cend())
+    {
+      found_key = true;
+      break;
+    }
+  }
+  return found_key;
+}
+
+bool SpeedScalingStateController::init_sensor_data()
+{
+  // loop in reverse order, this maintains the order of values at retrieval time
+  for (auto si = state_interfaces_.crbegin(); si != state_interfaces_.crend(); si++)
+  {
+    // initialize map if name is new
+    if (name_if_value_mapping_.count(si->get_name()) == 0)
+    {
+      name_if_value_mapping_[si->get_name()] = {};
+    }
+    // add interface name
+    name_if_value_mapping_[si->get_name()][si->get_interface_name()] = kUninitializedValue;
+  }
+
+  // filter state interfaces that have a speed scaling factor
+  // the rest will be ignored for this message
+  for (const auto& name_ifv : name_if_value_mapping_)
+  {
+    const auto& interfaces_and_values = name_ifv.second;
+    if (has_any_key(interfaces_and_values, { "speed_scaling_factor" }))
+    {
+      sensor_names_.push_back(name_ifv.first);
+    }
+  }
+
+  return true;
+}
+
+double get_value(const std::unordered_map<std::string, std::unordered_map<std::string, double>>& map,
+                 const std::string& name, const std::string& interface_name)
+{
+  const auto& interfaces_and_values = map.at(name);
+  const auto interface_and_value = interfaces_and_values.find(interface_name);
+  if (interface_and_value != interfaces_and_values.cend())
+  {
+    return interface_and_value->second;
+  }
+  else
+  {
+    return kUninitializedValue;
+  }
+}
+
+controller_interface::return_type SpeedScalingStateController::update()
+{
+  if (publish_rate_ > 0.0 && (node_->now() - last_publish_time_) > rclcpp::Duration(1.0 / publish_rate_, 0.0))
+  {
+    for (const auto& state_interface : state_interfaces_)
```

Comment:
Can be removed if only one state interface is configured

Reply:
Done. Simplified the whole controller